juju/juju
# Wireup the dqlite minimal app and unit services

This PR wires up the dqlite backed application and unit services to add and delete apps and units in dqlite to correspond with those in mongo state. The integration points are:
- bootstrap (add controller app and unit)
- deploy (add application api)
- add unit
- cleanup (remove app and unit)
- enable ha (add new controller units)
- migration model import

There's a little bit of churn in tests because of the helper AddApplication method now needing to take a mock for the dqlite service (since the real code dual writes the application data). This will all go away as we move away from mongo. What's here is just the minimum to integrate with the code which updates mongo state since it's all being rewritten.

During QA, an issue was discovered with migration - the service factory passed to the importer had been set up for the controller model, not the target model being migrated. This meant that the dqlite records for the machines, app, units were inserted into the controller model. So the apiserver root needs to be able to hand out a service factory getter to enable the model imported to update the correct dqlite database. This change is in the [second commit](https://github.com/juju/juju/pull/16887/commits/0b172b3896c49941012eb20e63c54df883e25019).

## QA steps

bootstrap
add model
deploy ubuntu -n 2
add-unit ubuntu
```
dqlite> select * from application;
53533bd4-038c-4148-8add-1d191a5007c2|ubuntu|0
dqlite> select * from unit;
c7835d21-2aed-4761-8c6f-20dfbc47f89d|ubuntu/0|53533bd4-038c-4148-8add-1d191a5007c2|eed19bd0-968d-453b-802d-d3a67fecceb7|0
4b975791-d3d2-4408-87f7-98311e6dbd75|ubuntu/1|53533bd4-038c-4148-8add-1d191a5007c2|24c4f0d4-fbf3-4717-875a-f607a04e0196|0
88054032-2658-4177-8a19-53740560eff8|ubuntu/2|53533bd4-038c-4148-8add-1d191a5007c2|6dc9175d-89ae-4e2b-87a9-222cfdf6050b|0
```

enable-ha
```
dqlite> select * from unit;
ac840fda-7ce0-43d0-8a2c-689868ab133e|controller/0|143992fc-232f-49e0-8904-d48de436983c|1b53a18a-a51f-4a21-88c8-88e98896f2c7|0
52a4a0ad-b1b2-424c-88bb-b09256e70cf2|controller/1|143992fc-232f-49e0-8904-d48de436983c|15d245a0-e7b4-4ded-88c5-936213464a7a|0
ef66de60-8509-4605-8cc5-6158c6a77241|controller/2|143992fc-232f-49e0-8904-d48de436983c|41bfa08c-f636-4dec-8ea9-ed5104daaed0|0
```

migrate model, check app and unit tables in the target model

delete unit and application and check rows are removed from dqlite

## Links

**Jira card:** JUJU-5406
